lib/chainpoint/hash.rb

```diff
@@ -0,0 +1,48 @@
+require 'digest'
+
+module Chainpoint
+  class Hash
+    NUM_SERVERS = 3
+
+    attr_reader :hash
+
+    def self.from_data(data)
+      new(Digest::SHA256.digest(data).unpack1('H*'))
+    end
+
+    def initialize(hash:, proof_handles: [])
+      @hash = hash
+      @proof_handles = proof_handles.map do |data|
+                         ProofHandle.new(uri: data[:uri], node_hash_id: data[:node_hash_id])
+                      end
+    end
+
+    def submit
+      @proof_handles = select_nodes(NUM_SERVERS).flat_map do |uri|
```

Comment:
@jaredmoody 

Does `Hash` send a request?
I thought `Hash` is a value object and other class as gateway or service should handle communication. 🤔 

```
Chainpoint.submit(hash)
Chainpoint.get_proof(hash)
```

Reply:
But the way is something like rails, so it might be appropriate. 🤔

Reply:
> Does Hash send a request?
> I thought Hash is a value object and other class as gateway or service should handle communication. 🤔

I'm treating `Hash` not just as a value object - but an object containing both data and behavior.  In this design, I've chosen to separate concerns over grouping similar behavior - e.g. making a network request.  Similar behavior could be extracted out via inheritance or mixin or a collaborator class though.

So, yes,`Hash` does send a request - to submit itself, and so does `Proof` to verify itself.  I've designed each object so that it alone contains the knowledge of how handle the things it's concerned with; `Hash` contains the logic of how it submits itself, including server selection.

Overall though do you like this direction?

Reply:
@jaredmoody 

No problem.
If that is not good design, someone will give some comments.